### PR TITLE
perf(dspark): fuse full-vocabulary greedy Markov reduction

### PR DESCRIPTION
**Pull request by voipmonitor**
## Purpose and behavior

DSpark greedy drafting repeatedly computes `(base_logits + markov_bias).argmax(-1)` over the full vocabulary. Fuse the addition and a two-stage argmax, and write directly into the caller-owned draft-token column. This removes the materialized sum and separate token copy without changing the Markov projection or truncating the vocabulary.

- Preserve source-dtype addition rounding, first-index ties, and first-NaN selection.
- Allocate FP32/int32 partial-reduction scratch once per proposer; reuse it during CUDA graph replay.
- Select the implementation only for full-vocabulary greedy proposals with matching input dtypes. Probabilistic, reduced-vocabulary, and gathered-top-k proposal paths retain their existing sampling implementations.
- Leave target weights/logits, rejection sampling, adaptive verification, and B12X unchanged. No additional configuration switch is required.

Status: **implemented**, with **qualified GPU operation parity** under the conditions below. A stable whole-model throughput gain remains **research-only**.

## Correctness validation

```bash
/opt/venv/bin/python -m pytest -q --confcutdir=tests/v1/worker \
  tests/v1/worker/test_dspark_greedy_markov.py
```

**30 passed in 7.37 s** in the serving image. Coverage: BF16/FP16/FP32; 1/4 rows; vocabulary sizes 1, 2047, 2048, 2049, and 129280; strided request/output tensors; vocabulary tails; ties, infinities and NaNs; input mutation; poisoned scratch; CUDA graph replay with unchanged allocation count; untouched neighboring output columns. These tests are self-contained; `--confcutdir` excludes parent fixtures requiring `tblib`, which is absent from the serving dependency set.

A diagnostic TP4 worker also compared every fused invocation against the materialized PyTorch operation on the **same live logits**, including graph replay. Three 1024-output-token requests and the prefill sequence completed without an assertion. That diagnostic extension is not part of the patch or serving image. Ruff, applicable commit hooks, and `git diff --check` passed for the source change.

Whole-completion bit equality is not established: repeated temperature-zero completions differ even within the unmodified control. No generated-code quality evaluation was performed; Sieve below is a coding-prompt speed probe.

## GPU microbenchmark

RTX PRO 6000 Blackwell Workstation, memory offset **+6000**. CUDA graphs execute seven full-vocabulary reductions at vocabulary size 129280. Both arms include addition, argmax, and output-column writes. After warmup, 41 samples alternate measurement order; output tokens match exactly.

| Dtype / requests | Materialized operation | Fused operation |
|---|---:|---:|
| BF16 / 1 | 135.82 µs | 13.85 µs |
| BF16 / 4 | 144.17 µs | 15.86 µs |
| FP32 / 1 | 134.79 µs | 14.84 µs |
| FP32 / 4 | 142.80 µs | 17.40 µs |

This saves about **0.12 ms per seven-position reduction sequence**, not 10× model throughput. The benchmark excludes the Markov projection itself.

## Serving observations and limits

Both arms used the same physical quartet: TP4/DCP1, native DeepSeek V4.1 Flash DSpark K7, greedy proposals, standard rejection, adaptive verification, V2, breakable `FULL_AND_PIECEWISE` graphs, B12X backends, mapped pinned-RAM Engram, batch budget 4096, max sequences 4, OMP8, NCCL16/2 MiB, no expert parallelism. Target sampling: temperature 1 / top-p 0.95. Memory offset **+6000**, graphics offset zero, automatic graphics clocks, 600 W limit; these are **not stock-clock results**.

C1/context-zero: five cells per arm, each requesting 15 s warmup and 30 s measurement. Sieve: one excluded warmup and five measured requests. Seeds were not paired across these stochastic probes.

| Five-run median | Control | Fused | Observed change |
|---|---:|---:|---:|
| C1 output tokens/s | 241.10 | 250.12 | +3.74% |
| C1 verifier steps/s | 90.67 | 91.53 | +0.94% |
| C1 emitted tokens/step | 2.653 | 2.735 | +3.10% |
| Sieve output tokens/s | 358.02 | 366.14 | +2.27% |
| Sieve min–max | 320.29–381.70 | 355.48–385.38 | Overlap |

Acceptance and adaptive verifier width change the work per step, so output differences do **not** establish an equivalent compute speedup. A fused-image restart measured 266.67 output tokens/s but 90.29 steps/s and 2.953 emitted tokens/step, reinforcing that limitation.

Cold unique 32768-token prefills used one output token and 30-second cells after warmup. Counters confirmed all 32768 tokens computed, with zero local/external cache hits:

| Process / measurement | Input tokens/s |
|---|---:|
| Control | 14793.57 |
| Control restart | 14757.88 |
| Fused, first process | 12832.12 |
| Fused, same first process repeated | 12997.00 |
| Fused, live-parity diagnostic process, profiler inactive | 14858.35 |
| Fused, uninstrumented restart before decode | 15004.55 |
| Fused, same restart after C1 and Sieve | 14810.74 |

**The first-process prefill slowdown remains unexplained.** Ordinary fused-image launches used identical source, flags, clocks, and cache directory. Restart results match the control, but do not qualify first-boot reliability or prove the cause. This PR does not claim a prefill fix or a regression-free release.

## Scope, provenance, and related work

The PR changes only the DSpark sampler and its test. Both serving arms include the separate B12X TP-padding correction [b12x#361](https://github.com/local-inference-lab/b12x/pull/361); it is not part of this optimization.

Open fork/upstream PRs were checked before submission. [vLLM #736](https://github.com/local-inference-lab/vllm/pull/736) concerns MoE/Engram execution, not sampling. [Upstream #50737](https://github.com/vllm-project/vllm/pull/50737) fuses bias addition into the Markov GEMM and changes logit layout; this PR instead preserves the projection and fuses the full-vocabulary reduction. They are related alternatives for removing the addition pass, not independently additive speedup claims. [Upstream #54433](https://github.com/vllm-project/vllm/pull/54433) restricts proposal candidates; this PR does not. [Upstream #50843](https://github.com/vllm-project/vllm/pull/50843) concerns other sampler tail bounds; this reduction masks tail indices and tests partial tiles and all-negative-infinity rows.

<details>
<summary>Immutable source and image identities</summary>

- Base JJ commit: `377c93be7e74619c629a3177292f49deac5639fd`.
- PR commit: `9d0c33e4227db6fe09b8b838a84a47b40e970d57`.
- Tested/PR source tree: `d59cf27e8f50b911a708d9fadc9c98e9045305ac`.
- Tested image records commit `956cd4b67edaee8b408bd5c18079a58e95ef9832`; its tree is identical to the PR tree. Only commit metadata differs, including the AI co-author trailer.
- Control local image ID: `sha256:da403ee8ceff7b510aeae1a04ca751478b2413099cfde4b45b99b96397ea0bd4`.
- Fused local image ID: `sha256:4dc417893f9f32b6fdeceba3ccfb23e6d18c74ff3446074c0f5285f93db6c264`.
- Both arms use B12X `fcecbfe605bd3e1155efd2df2d983d012e68f987`; native libraries are unchanged. The images are local test artifacts, not published registry releases.

</details>

AI assistance was used for implementation, profiling, validation, and this description. Human maintainer review is requested; this description does not assert that such review has already occurred.

## Summary by CodeRabbit

- **New Features**
  - Added greedy Markov sampling for speculative decoding.
  - Supports exact vocabulary-wide selection across multiple data types and request sizes.
  - Preserves expected handling of ties, special values, and strided outputs.
  - Supports CUDA graph replay without intermediate allocations for improved efficiency.
  - Reports invalid vocabulary dimensions and insufficient workspace before GPU execution.

- **Tests**
  - Added CUDA coverage for eager execution and graph replay, including input mutation, output correctness, sentinel preservation, and stable memory usage.
  - Added validation coverage for invalid input dimensions and workspace sizes.

### CODE REVIEW

**Comment by voipmonitor** [comment]
@coderabbitai review

File: vllm/v1/worker/gpu/spec_decode/dspark/speculator.py [hunk 1]
```diff
@@ -85,6 +89,17 @@ def __init__(self, vllm_config: VllmConfig, device: torch.device):
...
+            getattr(self.draft_model_config.hf_config, "draft_vocab_size", None)
+            or self.draft_model_config.hf_config.vocab_size
+        )
+        shape = scratch_shape(self.max_num_reqs, draft_vocab)
+        self._greedy_partial_values = torch.empty(
+            shape, dtype=torch.float32, device=device
+        )
+        self._greedy_partial_indices = torch.empty(
+            shape, dtype=torch.int32, device=device
+        )
```
[the remark this answers is not in the document]
Comment (voipmonitor) [reply vllm/v1/worker/gpu/spec_decode/dspark/speculator.py]: Confirmed the invalid-shape failure mode. Commit 17db68aec6b bounds the proposer allocation by both configured vocabulary widths and rejects unequal actual base/bias shapes or insufficient scratch before either kernel launches. No vocabulary truncation or fallback is introduced; supported shapes execute unchanged GPU kernels. The production DS4.1 checkpoint has matching 129280-column heads. Validation: 34 tests passed on RTX PRO 6000 (8.53 s), covering both mismatch directions, undersized scratch, BF16/FP16/FP32 exact parity and graph replay. Applicable pre-commit checks passed. The clarified Args docstring is included.

**Comment by voipmonitor** [comment]
Qualified the source-locked R36 image on four RTX PRO 6000 Blackwell Workstation GPUs at TP4/DCP1, DSpark K7 greedy proposals with standard rejection and adaptive verification. Both RAM and SSD Engram modes used the same physical GPUs, VRAM +6000, temperature 1, top-p 0.95 and a 4,096-token scheduler budget.

- First seeded processed-FP32 request: passed with the strict JIT monitor enabled.
- Arithmetic and image-recognition smoke: passed in both modes.
- Target, draft and piecewise graph capture: completed in both modes.
- Exact image: 106 GPU sampler/warmup tests passed; portable recipe: 200 CPU tests passed.
- RAM: 32K uncached prefill 14,279 before / 13,988 after decode; C1 median 260.8 tok/s and 97.35 verifier steps/s; five Sieve requests median 391.7 tok/s.
- SSD: 32K uncached prefill 14,062 / 14,030; C1 median 244.3 tok/s and 90.43 verifier steps/s; Sieve median 361.4 tok/s.

The two-layer registry image is `localinferencelab/vllm:jovian-judgement-community-20260912-r36`. [Runbook, immutable identity and numeric samples](https://github.com/local-inference-lab/rtx6kpro/blob/master/models/deepseek-v4.1-flash.md).

This is a bounded serving qualification, not a claim of deterministic whole-model output or long-context production quality. #738 has exact per-operation token parity; a stable whole-model speedup is not established. #740 changes startup specialization coverage, not runtime sampling math, and is not claimed to fix the separately observed intermittent prefill slowdown.
